add tls options for Agent

ahalimx86 opened the pull request:
Add tls option for Infra Agent. 
Server side option will follow in subsequent commits

ahalimx86 committed:
add infra-agent tls specific CLI options

add infra Agent TLS specic CLI and config file options for Client Cert,
Client Key, CA Cert, insecure and mtls flags.

Signed-off-by: Abdul Halim <abdul.halim@intel.com>

ahalimx86 committed:
update agent's grpc client conn with tls options

Update all gRPC client connection to Infra Manager with TLS option when
TLS cofigs are given.

Signed-off-by: Abdul Halim <abdul.halim@intel.com>

venkateshpnv commented on infraagent/agent/cmd/agent/rootcmd.go:
```diff
@@ -87,6 +92,12 @@ func init() {
+	rootCmd.PersistentFlags().StringVar(&config.clientCert, "client-cert", types.AgentDefaultClientCert, "TLS Client cert file for mTLS")
+	rootCmd.PersistentFlags().StringVar(&config.clientKey, "client-key", types.AgentDefaultClientKey, "TLS Client key file for mTLS")
+	rootCmd.PersistentFlags().StringVar(&config.caCert, "ca-cert", types.AgentDefaultCACert, "TLS Client CA Cert file")
+
```
I guess we also need the server CA to verify the server's certificate. Do we need client CA? It is required at the server end right?

ahalimx86 replied:
So, Agent will not run  it's server with TLS as calico does not support it.
In Infra Manager, yes we will need both client-cert, client-key, server-cert, server-key and ca-cert.
the client-cert, client-key is for communicating to P4d and the server-cert, server-key for exposing service to Infra-Agent.
The ca-cert must be issued by same issuer for Infra agent and server.

ahalimx86 committed:
infra agent run in insecure mode

Until the full mTLS support is complete run Infra Agent in insecure mode
to that current deployment mode still works. Infra agent configuration
is provided via existing 'infraagent-config' configMap.

Signed-off-by: Abdul Halim <abdul.halim@intel.com>

venkateshpnv commented on pkg/utils/tls.go:
```diff
@@ -0,0 +1,93 @@
+		return insecure.NewCredentials(), nil
+	}
+
+	if viper.GetBool("mlts") {
```
I think this is "mtls".

ahalimx86 replied:
fixed

ahalimx86 committed:
add create tls secrets with gen-cert scripts

Adding make target 'tls-secrets' to create required secret volume mounts
for K8s deployment with TLS authentication. We use openssl based scripts
to generate self-signed certs and keys for Infra Agent and Manager.

Co-authored-by: Venkatesh Petla <neelakanta.venkatesh.petla@intel.com>

Signed-off-by: Abdul Halim <abdul.halim@intel.com>

ahalimx86 committed:
update Agent and Manager DS spec with secret vol

Update Agent and Manager DaemonSet specs with TLS secret volumes added.

Signed-off-by: Abdul Halim <abdul.halim@intel.com>

ahalimx86 committed:
fix typo in utils/tls.go

Signed-off-by: Abdul Halim <abdul.halim@intel.com>

dhruvpt commented on Makefile:
```diff
@@ -1,6 +1,11 @@
 IMAGE_VERSION?=latest
 KUBECONFIG?=$(HOME)/.kube/config
+KUBENAMESPACE?=kube-system
+K8S_SECRETE_MGR_SERVER?=manager-server-secret
```
Check spelling - SECRETE

dhruvpt commented on Makefile:
```diff
@@ -113,6 +118,21 @@ create-kubeconfig-cm:
+
+.PHONY : gen-certs
+gen-certs:
+	 ./hack/tls/gen_certs.sh
```
We should keep the generation of keys and certificates as decoupled from k8s as possible. For that, the makefile should not be invoking scripts internally to automatically generated these files. Instead, the makefile should assume that certificates have been pre-created at the required location and so, it can go and grab them for usage. The script can be provided purely as a standalone reference so that user can use it or have their own way of generating these files with their choice of encryption algorithm and keys/certificates, etc. as we discussed in the context of SDLe.

ahalimx86 replied:
The secrets with certs & keys are still optional and a user can always provide their own files and create the K8s secrets manually if they wish. These make targets are not pre-requisites for deploy but the secrets are. It does not matter if they create them with provided make target or manually.

dhruvpt commented on deploy/infraagent-configmap.yaml:
```diff
@@ -6,3 +6,4 @@ metadata:
 data:
   infraagent.yaml: |
     interfaceType: tap
+    insecure: true
```
Are we defaulting to insecure mode? I don't think that's acceptable from SDLe perspective.

ahalimx86 replied:
As mentioned in above comment this is temporary config. (it was described in the [commit message](https://github.com/ipdk-io/k8s-infra-offload/pull/37/commits/164ad0f003e5a759a48694f734df52d1e8356c24) as well )

dhruvpt replied:
ok, I missed the commit message. As long as we don't make a release with this default setting, we should be okay.

venkateshpnv commented on deploy/infraagent-configmap.yaml:
```diff
@@ -6,3 +6,4 @@ metadata:
 data:
   infraagent.yaml: |
     interfaceType: tap
+    insecure: true
```
Our understanding is that the agent and manager will communicate via mtls. And the insecure flag is set to false by default. Do we need to it to be explicitly set to true here?

ahalimx86 replied:
I have added this as temporary config so that deployment of Agent still works while mTLS is not fully implemented. We can remove or change this when TLS is fully functional.

dhruvpt commented on hack/tls/gen_certs.sh:
```diff
@@ -0,0 +1,101 @@
+#!/bin/bash
```
This file, serving as reference, should be moved at the top level in the Scripts directory.
User then should be able to modify any of the parameters as needed.

ahalimx86 replied:
I dont have any preference. This directory is also meant for keeping scripts etc.

dhruvpt replied:
Yeah, we can leave this directory for CI/CD related scripts. And use the Scripts directory at the root for deployment related scripts.

dhruvpt commented on hack/tls/openssl.cnf:
```diff
@@ -0,0 +1,31 @@
+[req]
```
This can also be moved at top level along with other config files for user to modify as per the deployment needs.
BTW, why is the directory name "hack"?

ahalimx86 replied:
It's just a common practice in Go or K8s related project.

_"The [hack](https://github.com/kubernetes/kubernetes/blob/master/hack) directory contains many scripts that ensure continuous development of kubernetes, enhance the robustness of the code, improve development efficiency, etc. The explanations and descriptions of these scripts are helpful for contributors. For details, refer to the following guidelines."_

Source: [kubernetes/hack](https://github.com/kubernetes/kubernetes/tree/master/hack#overview)

dhruvpt replied:
Thanks for the pointer! But going through it, it seems that's how Kubernetes has used the hack directory, but it is not a convention to be followed. Moreover, Kubernetes uses this directory to store all scripts related to building and testing for product stability and scripts for common updates -- for efficiency and uniformity.
For us, we already have a scripts directory in the root folder, having create-interfaces script to be used for deployment. Similar scripts, including the one to generate certificates, that are meant for deployment, can be placed in it.

ahalimx86 replied:
moved gen-certs under scripts dir.

ahalimx86 committed:
move gen-certs scripts under scripts dir

Moving gen-certs scripts inside scripts dir based on review comments.

Signed-off-by: Abdul Halim <abdul.halim@intel.com>

nupurjai commented on deploy/inframanager-daemonset.yaml:
```diff
@@ -38,6 +38,10 @@ spec:
         - name: db
           mountPath: /var/lib/cni/inframanager
+        - name: client-certs
+          mountPath: /tmp/inframanager/client/
```
/tmp/inframanager/cert/client/  -?

nupurjai commented on pkg/utils/tls.go:
```diff
@@ -0,0 +1,93 @@
+	}
+
+	if viper.GetBool("mtls") {
+		return getClientMTLSCredentials()
```
Instead of bool, if we have enums - 0 insecure 1 TLS 2 mTLS

ahalimx86 replied:
As discussed over call, this can be reviewed again as future improvement.

nupurjai reviewed (commented):
Loos like the insecure mode has a problem.

venkateshpnv commented on pkg/utils/utils.go:
```diff
@@ -531,7 +530,7 @@ func getHealthServerResponse(conn *grpc.ClientConn) (*healthpb.HealthCheckRespon
 // CheckGrpcServerStatus will check gRPC server status using gRPC health check
 func CheckGrpcServerStatus(target string, log *log.Entry, grpcDial grpcDialType) (bool, error) {
-	conn, err := grpcDial(target, grpc.WithTransportCredentials(insecure.NewCredentials()))
+	conn, err := grpcDial(target)
```
I think you should call getCredentials() here too.
  
credentials, err := getCredentialFunc()
if err != nil {
        s.log.WithError(err).Error("error getting gRPC client credentials to connect to backend")
        return out, nil
    }
    conn, err := grpcDial(managerAddr, grpc.WithTransportCredentials(credentials))

ahalimx86 replied:
missed this one here. Added fix in latest commit

nupurjai commented on infraagent/agent/cmd/agent/rootcmd.go:
```diff
@@ -87,6 +92,12 @@ func init() {
 	rootCmd.PersistentFlags().StringVar(&config.cfgFile, "config", "/etc/infra/infraagent.yaml", "config file")
 	rootCmd.PersistentFlags().StringVar(&config.tapPrefix, "tapPrefix", types.TapInterfacePrefix, "Host TAP interface prefix for TAP interface type")
+	rootCmd.PersistentFlags().BoolVar(&config.insecure, "insecure", false, "use insecure mode for internal communication with backend")
+	rootCmd.PersistentFlags().BoolVar(&config.mtls, "mtls", true, "use mTLS for internal communication with backend")
```
Two flags - insecure
and mTLS

Mode - insecure, mTLS, TLS

ahalimx86 replied:
As discussed over call, this can be reviewed again as future improvement.

venkateshpnv commented on Makefile:
```diff
@@ -113,6 +119,21 @@ create-kubeconfig-cm:
 delete-kubeconfig-cm:
 	kubectl -n kube-system delete configmap $(KUBECONFIG_CM)
 
+.PHONY : tls-secret
```
Is it "tls-secrets"  instead of "tls-secret" ?

ahalimx86 replied:
It's a typo. Should be `tls-secrets`

ahalimx86 committed:
fix check_manager grpc dialer with TLS

Update check manager with grpc dialer with TLS credentials that was
missing.

Signed-off-by: Abdul Halim <abdul.halim@intel.com>

ahalimx86 committed:
fix typo in Makefile

fix a typo in Makefile

Signed-off-by: Abdul Halim <abdul.halim@intel.com>

ahalimx86 committed:
fix pkg/services unit tests

fix pkg/services unit test for grpc with fake credentials.

Signed-off-by: Abdul Halim <abdul.halim@intel.com>

ahalimx86 committed:
fix healthcheck tls dialer for CNI server

The CNI server health check is not running with TLS therefore client
connetion from health check needs to use default credentials.

Signed-off-by: Abdul Halim <abdul.halim@intel.com>

ahalimx86 committed:
add tls secret volumes for check-for-manager init container

add tls secret volumes for check-for-manager init container which was
missing.

Signed-off-by: Abdul Halim <abdul.halim@intel.com>

ahalimx86 committed:
update tls certs default location name

update tls certs default location path name as suggested in PR comment.

Signed-off-by: Abdul Halim <abdul.halim@intel.com>
